docs/proposals/001-k8s-status.md

```diff
@@ -0,0 +1,433 @@
+below, or serve as reference for future proposals.
+-->
+
+### `message` field
+
+The `message` field is the only information that will be presented to the user.
+We should be explicit that this field is only meant to be informative, users
+should not try to programmatically parse it to extract meaningful information
+about the feature.
+
+It is out-of-scope for this proposal to provide any sort of _structured_
+fields into the output of `k8s status` command.
+
+### `version` field
+
+This field is set to the application version of the feature (e.g. `v1.15.2`).
+In the future, this may be used to detect when we _can_ or _should_ upgrade
+the version of features (e.g. a newer Canonical Kubernetes version would come
+with Cilium v1.16.2), we could add a similar `upgradeable-to` field.
+
+This is out of scope for this proposal.
+
+### `enabled` field
+
+For the output of `k8s status`, we only care that the feature itself has been
+successfully deployed on the cluster. When this field is true, it means that any
+manifests required were successfully deployed on the cluster.
+
+In the future, we might want to extend this with a `ready: true/false` field,
+indicating that the feature is ready and/or available and/or healthy.
+
+This is out of scope for this proposal.
+
+# Implementation Details
+
+## API Changes
+<!--
+This section MUST mention any changes to the k8sd API, or any additional API
+endpoints (and messages) that are required for this proposal.
+
+Unless there is a particularly strong reason, it is preferable to add new v2/v3
+APIs endpoints instead of breaking the existing APIs, such that API clients are
+not affected.
+-->
+
+Add a new type in [api/v1/types.go]
+
+```go
+type FeatureStatus struct {
+    Enabled bool
+    Message string
+    Version string
+    Timestamp time.Time
+}
+```
+
+Extend [apiv1.ClusterStatus] with fields for the status of individual features.
+
+```go
+type ClusterStatus struct {
+    // ...
+    Network FeatureStatus
+    DNS FeatureStatus
+}
+```
+
+Extend [pkg/k8sd/types] with the internal `FeatureStatus` struct (this should be
+a separate struct, such that internal types and API can evolve separately).
+
+```go
+type FeatureStatus struct {
+    Enabled bool
+    Message string
+    Version string
+    Timestamp time.Time
+}
+```
+
+Two helper methods should also be added to convert between internal<->API types:
+
+```go
+func (f FeatureStatus) ToAPI() apiv1.FeatureStatus
+func FeatureStatusFromAPI(apiv1.FeatureStatus) FeatureStatus
+```
+
+## CLI Changes
+<!--
+This section MUST mention any changes to the k8s CLI, e.g. new arguments,
+different outputs.
+-->
+
+These changes will affect the output of the `k8s status` command, as explained
+in the section above.
+
+## Database Changes
+<!--
+This section MUST mention any changes required in the k8sd database schema or
+internal types.
+-->
+
+We need to define a new table (changes go in [pkg/k8sd/database]):
```

Comment:
I'm a bit confused about the database that we're adding this table to. AFAIK, there's `k8sd dqlite` which stores information about the nodes and does not really know anything about the actual k8s implementation on top of it (which nodes are master/worker, what is installed on each node or the cluster etc.) and there's this `k8s dqlite` which stores the actual cluster info known to k8s. we're changing the `k8sd/database` here which seems a bit odd to me. shouldn't this info be stored in the k8s database rather than the "node manager guy (k8sd)" database?

p.s. I can see that there is a `cluster_configs` table in the k8sd dqlite, but again, is k8sd (which afaik is a wrapper around MicroCluster) really supposed to know about the details of the actual k8s on top of it?

Reply:
So, `k8sd` is the orchestrator layer, exactly the place where we want this data to be stored. the kubernetes cluster itself cannot see this data, or interact with it.